File: commands/disallow.py
import discord

from function.file import load_data, save_data
from dotenv import load_dotenv
import os


load_dotenv()

YOUR_ID = int(os.getenv("YOUR_ID"))
async def handle_disallow(ctx, user: discord.Member):
    guild_id = ctx.guild.id

    allowed = load_data(guild_id, "allowed.json")

    if not ctx.user.server_permissions.administrator or ctx.user.id not in [ctx.guild.owner_id, YOUR_ID, ]:
        await ctx.response.send_message("You do not have permission to use this command!")
        return

    if str(user.id) not in allowed:
        await ctx.response.send_message(f"<@{user.id}> is not on allowed list.", ephemeral=True)
        return

    del allowed[str(user.id)]
    save_data(guild_id, "allowed.json", allowed)
    await ctx.response.send_message(f"<@{user.id}> has been removed from the special privileges list.")

File: src/commands/disallow.ts
// src/commands/disallow.ts

import { loadData, saveData } from "@/utils/file.js";
import { createLogger } from "@/utils/log.js";
import { isMediaAdmin } from "@/utils/permissions.js";
import { SlashCommandBuilder } from "@discordjs/builders";
import { InteractionContextType } from "discord-api-types/v10";
import { ChatInputCommandInteraction, MessageFlags } from "discord.js";

const log = createLogger("cmd/disallow");

/** Slash command definition for `/disallow` - revoke bot-admin privileges. */
export const data = new SlashCommandBuilder()
  .setName("disallow")
  .setDescription("🔒 Take someone's bot-admin privileges away")
  .addUserOption((opt) =>
    opt.setName("user").setDescription("Member to disallow").setRequired(true),
  )
  .setContexts(InteractionContextType.Guild);

/**
 * Executes `/disallow`: removes a member from the guild's allowed list.
 * @param interaction - The command interaction context.
 * @returns A promise that resolves when the reply has been sent.
 */
export async function execute(interaction: ChatInputCommandInteraction): Promise<void> {
  if (!interaction.inGuild()) {
    await interaction.reply({ content: "Use in a server.", flags: MessageFlags.Ephemeral });
    return;
  }
  if (!isMediaAdmin(interaction)) {
    await interaction.reply({
      content: "❌ You're not allowed to run this.",
      flags: MessageFlags.Ephemeral,
    });
    return;
  }

  const user = interaction.options.getUser("user", true);
  const guildId = interaction.guildId;
  const config = loadData<{ allowed?: string[] }>(guildId, "allowed.json", { soft: true });
  const allowed = config.allowed ?? [];

  if (!allowed.includes(user.id)) {
    await interaction.reply({
      content: `${user} was not on the allowed list.`,
      flags: MessageFlags.Ephemeral,
      allowedMentions: { parse: [] },
    });
    return;
  }

  saveData(guildId, "allowed.json", { ...config, allowed: allowed.filter((id) => id !== user.id) });
  log.info("user disallowed", { guildId, userId: user.id, by: interaction.user.id });
  await interaction.reply({
    content: `✅ ${user} no longer has special privileges.`,
    allowedMentions: { parse: [] },
  });
}
